File: spec-traits/src/main/java/bsp/traits/DataTrait.java
package bsp.traits;

import software.amazon.smithy.model.node.Node;
import software.amazon.smithy.model.node.ObjectNode;
import software.amazon.smithy.model.shapes.ShapeId;
import software.amazon.smithy.model.traits.AbstractTrait;
import software.amazon.smithy.model.traits.AnnotationTrait;

public class DataTrait extends AnnotationTrait {

    public static ShapeId ID = ShapeId.from("jsonrpc#data");

    public DataTrait(ObjectNode node) {
        super(ID, node);
    }

    public DataTrait() {
        super(ID, Node.objectNode());
    }

    public static final class Provider extends AbstractTrait.Provider {
        public Provider() {
            super(ID);
        }

        @Override
        public DataTrait createTrait(ShapeId target, Node node) {
            return new DataTrait(node.expectObjectNode());
        }
    }
}

File: spec-traits/src/main/kotlin/traits/DataTrait.kt
package traits

import software.amazon.smithy.model.node.ObjectNode
import software.amazon.smithy.model.shapes.ShapeId
import software.amazon.smithy.model.traits.AnnotationTrait

class DataTrait(node: ObjectNode) : AnnotationTrait(ID, node) {

    class Provider : AnnotationTrait.Provider<DataTrait>(ID, ::DataTrait)

    companion object {
        val ID: ShapeId = ShapeId.from("traits#data")
    }
}